mqtt/mqtt-broker/src/sys_topic.rs

```diff
@@ -0,0 +1,389 @@
+        let connections = sessions
+            .iter()
+            .filter_map(|(client_id, session)| match session {
+                Session::Transient(_) => Some(client_id),
+                Session::Persistent(_) => Some(client_id),
+                _ => None,
+            })
+            .collect();
+
+        Self::Connections(connections)
+    }
+
+    pub fn new_session(sessions: &'a HashMap<ClientId, Session>) -> Self {
+        let sessions = sessions
+            .iter()
+            .filter_map(|(client_id, session)| match session {
+                Session::Transient(_) => Some(client_id),
+                Session::Persistent(_) => Some(client_id),
+                Session::Offline(_) => Some(client_id),
+                _ => None,
+            })
+            .collect();
+
+        Self::Sessions(sessions)
+    }
+}
+
+impl<'a> TryFrom<StateChange<'a>> for proto::Publication {
+    type Error = Error;
+
+    fn try_from(state: StateChange<'a>) -> Result<Self, Error> {
+        Ok(match state {
+            StateChange::Subscriptions(client_id, subscriptions) => {
+                let payload = if let Some(subscriptions) = subscriptions {
+                    serde_json::to_string(&subscriptions)
+                        .map_err(|_| Error::NotifyError)?
+                        .into()
+                } else {
+                    "".into()
+                };
+
+                proto::Publication {
+                    topic_name: format!("$edgehub/subscriptions/{}", client_id),
+                    qos: STATE_CHANGE_QOS,
+                    retain: true,
+                    payload,
+                }
+            }
+            StateChange::Connections(connections) => proto::Publication {
+                topic_name: "$edgehub/connected".to_owned(),
+                qos: STATE_CHANGE_QOS,
+                retain: true,
+                payload: serde_json::to_string(&connections)
+                    .map_err(|_| Error::NotifyError)?
+                    .into(),
+            },
+            StateChange::Sessions(sessions) => proto::Publication {
+                topic_name: "$edgehub/sessions".to_owned(),
+                qos: STATE_CHANGE_QOS,
+                retain: true,
+                payload: serde_json::to_string(&sessions)
+                    .map_err(|_| Error::NotifyError)?
+                    .into(),
+            },
+        })
+    }
+}
+
+#[cfg(test)]
+pub(crate) mod tests {
+    use super::*;
+    use crate::broker::tests::*;
+    use crate::session::*;
+    use crate::subscription::*;
+    use crate::*;
+    use std::collections::{HashMap, VecDeque};
+    use std::convert::TryInto;
+    use std::iter::FromIterator;
+    use std::str::FromStr;
+    #[test]
+    fn test_subscriptions() {
+        let expected_id: ClientId = "Session".into();
+
+        // Session with no subscriptions
+        let session = make_session(expected_id.as_str(), Vec::<&str>::new(), true);
```

Comment:
```suggestion
        let session = make_session(expected_id.as_str(), Vec::new(), true);
```

Reply:
Lets put type only in places where it is really required. It doesn't add clarity but makes code harder to read

Reply:
```
error[E0282]: type annotations needed
   --> mqtt-broker\src\sys_topic.rs:116:23
    |
116 |         let session = make_session(expected_id.as_str(), Vec::new(), true);
    |                       ^^^^^^^^^^^^ cannot infer type for type parameter `S` declared on the 
function `make_session`
```

Reply:
This is b/c make_session takes an Iterator of AsRef<str>. I wanted it to take either &str or Strings

Reply:
Yeah, I think you may want to use `Into<String>` in the make_session method instead.